query-parser/src/parse.rs

```diff
@@ -70,6 +69,130 @@ macro_rules! matches_plain_symbol {
     }
 }
 
+trait FromValue<T> {
```

Comment:
`TryFrom` is unstable, no?

But I opted for `Option` because that's also the pattern `satisfy_map` uses. There isn't really a useful error class we could include: we just want to encode that this value wasn't a suitable `PatternNonValuePlace`. That's also why I kept this trait private: it's almost a stylistic utility for the parser, turning a collection of functions like `value_to_pattern_non_value_place` into a common method name.